### DISCUSSION TITLE
Importance of two way data binding between URL Params and UI

### DISCUSSION THREAD

**Discussion by kevscript** [vercel/next.js#58803] [Help]
Hey, I'm looking for some clarifications about the way I should handle the relationship between URLParams and the UI of my app.

<hr />

- Firstly, I've been learning how to implement filtering options through URL Params with [this official demo repo](https://github.com/vercel/nextjs-postgres-nextauth-tailwindcss-template/tree/main). Its running [here](https://next-admin-dash.vercel.app/).

If we update the value of the search input, the changes will reflect through `searchParams` on `q=[search_input_value]`, and the component will refetch the data based on the new filtering query. 

But when we update `searchParams.q` either directly `http://route.com/?q=foo` or redirect via a `<Link href={'/?q=foo'} />`, we can notice a visual mismatch between the URL (+fetched data) and the empty search input.

**Shouldn't the input also reflect the current value filtering the data?** 
A typical user will mostly rely on that search functionality to understand how the data he sees is being processed. An empty input suggest an unfiltered list, but the reality is different. Is UX and visual fidelity trivial in that case?

<hr />

- Secondly, I've been working on a similar use case, but implemented with a mandatory default filtering value.

Lets say I want to display a filterable list of `historicalEvents`, but I never want to fetch them all at once. I want them to be pre-filtered at least by `year`.  If the route is being called without a valid `year` param, how should the page behave? 

- Should I render an error if `searchParams.year` is `undefined` or implicitly fetch my data with a conditional default `year`?
- What if its not `undefined` but simply not a valid `year` value? Error out or conditional default?
- Should I fetch the data with a default `year` without caring about the empty/falsy URL state?
- Should I rerender the route with valid default URL Params before fetching my data?

It feels like there is so many moving pieces between URL Params, validation, fetching, rendering. Add localStorage or cookies for some data persistence and I don't know where to draw the lines anymore.

**Comment by nicitaacom** [comment]
> Shouldn't the input also reflect the current value filtering the data?

So as I understood you want to create searching functionality for your project (to search for products by title and subTitle for example)
In this case (should input have value the same as in URL) **depends on your requirements** - usually user will not change directly search param to find something

>  via a <Link href={'/?q=foo'} />

if you have some categories its fine that you have `value=""` for input (otherwise describe your specific use-case for clicking on link that pusht you to url with search params)

> Is UX and visual fidelity trivial in that case?

Depends on your requirements - its nothing wrong if you keep input with `value=""` or keep input value the same as in search params
If you looking for impression - I would recommend you 'don't explore your bike' and check how others do - usually its default UX

> Should I render an error if searchParams.year is undefined or implicitly fetch my data with a conditional default year?

As I understand context - you should don't let user to cause that error - if user may cause that error manually somehow (e.g change URL in browser searchbar) - in that case you say user in UI smth like 'now put valid year in search params'

> What if its not undefined but simply not a valid year value? Error out or conditional default?

The same - you should create functionality in your app to don't let user to cause some error

> Should I fetch the data with a default year without caring about the empty/falsy URL state?

When?
during SSR?
or when user click on button or submit form?

> Should I rerender the route with valid default URL Params before fetching my data?

This question I didn't understood - remember that you ask somebody who know nothing about project
So add clear question (or its just my understanding of what you wrote and somebody may help you better)

> It feels like there is so many moving pieces between URL Params, validation, fetching, rendering. Add localStorage or cookies for some data persistence and I don't know where to draw the lines anymore.

As I understood its question

> Add localStorage or cookies

Check video on YouTube about that

In short
use cookies if you need to get access to that cookis on server
use localstorage when you don't need to get access to data you save on server (to fetch some data based on data that you save in localstorage)

In general I would recommend you to ask ChatGPT about some questions - create some project clone to undestand best practices and how should work something

**Reply by kevscript** [reply]
> > Should I fetch the data with a default year without caring about the empty/falsy URL state?
> 
> When? during SSR? or when user click on button or submit form?
>
> > Should I rerender the route with valid default URL Params before fetching my data?
>
> This question I didn't understood - remember that you ask somebody who know nothing about project
> So add clear question (or its just my understanding of what you wrote and somebody may help you better)

```tsx
import { redirect } from 'next/navigation'
import EventsList from '@/components/EventsList'
import EventFilterBox from '@/components/EventFilterBox'
import isValid from 'validation-library'

const defaultYear = "2023"

const Page = async ({ searchParams }) => {
  // current URL: historicalEvents.com/?year=imNotaYear

  // year is not valid -> searchParams.year === "imNotaYear"
  const isValidYear = isValid("year", searchParams.year)

  // knowing the year is not valid, 

  // should I redirect to an URL with valid defaults before fetching
  !isValidYear && redirect(`/?year=${defaultYear}`)

  // or

  // should I fetch data directly with a default value
  const historicalEvents = await fetchEvents({ year: isValidYear ? searchParams.year : defaultYear /* 2023 */ })
  // while keeping the dirty URL "/?year=imNotaYear"

  // or 

  // create an error state to notify the user
  const errorMessage = isValidYear ? null : `${searchParams.year} is not a valid year.` 


  return {
    <>
        {!isValidYear ? (
            <>
               <span>{errorMessage}</span>
               <Link href={`/?year=${defaultYear}`}>
                    <button>Retry</button>
               </Link>
            </>
        ) : (
            <>
               <EventFilterBox />
               <EventsList events={historicalEvents} />
            </> 
        )}
    </>
  }
}
```

Basicly, if my URL is supposed to be the only source of truth, is it better to:
- keep it valid at all times and truely represent the state of my page before executing crucial steps like fetching or rendering.
- rely on valid default values that would "take-over" a faulty URL state if necessary.
- implement an error flow to make the user aware of the problem.
